File: framework/php/src/Sse.php
<?php
declare(strict_types=1);

namespace BetterPortal\Protocol;

/** UTF-8 SSE framing only; hosts own validation, subscriptions, flushing and cancellation. */
final class Sse
{
    /**
     * Encode one message, preserving empty data and Unicode line separators.
     * Data is already serialized text, not an unvalidated application value.
     * @throws \InvalidArgumentException for invalid UTF-8, unsafe fields or exceeded limits.
     */
    public static function encode(string $data, ?string $event = null, ?string $id = null, ?int $retry = null, int $maxDataBytes = 1048576): string
    {
        if ($maxDataBytes < 1 || strlen($data) > $maxDataBytes || preg_match('//u', $data) !== 1) {
            throw new \InvalidArgumentException('Invalid SSE data or byte limit exceeded');
        }
        foreach ([$event, $id] as $value) {
            if ($value !== null && (strlen($value) > 1024 || strpbrk($value, "\r\n\0") !== false || preg_match('//u', $value) !== 1)) {
                throw new \InvalidArgumentException('Invalid SSE event name or ID');
            }
        }
        if ($retry !== null && $retry < 0) {
            throw new \InvalidArgumentException('SSE retry must be nonnegative');
        }
        $fields = $event === null ? [] : ['event: ' . $event];
        foreach (explode("\n", str_replace(["\r\n", "\r"], "\n", $data)) as $line) {
            $fields[] = 'data: ' . $line;
        }
        if ($id !== null) {
            $fields[] = 'id: ' . $id;
        }
        if ($retry !== null) {
            $fields[] = 'retry: ' . $retry;
        }
        return implode("\n", $fields) . "\n\n";
    }
}

File: framework/go/sse.go
package protocol

import (
	"fmt"
	"strconv"
	"strings"
	"unicode/utf8"
)

// EventOptions controls framing only; nil fields are omitted and empty IDs reset
// the browser's last event ID. Zero MaxDataBytes defaults to 1 MiB.
type EventOptions struct {
	Event        *string
	ID           *string
	Retry        *int64
	MaxDataBytes int
}

// EncodeEvent frames already serialized UTF-8 data. Hosts own AnyVali validation,
// scope, authorization, subscriptions, backpressure, flushing and cancellation.
func EncodeEvent(data string, options EventOptions) ([]byte, error) {
	limit := options.MaxDataBytes
	if limit == 0 {
		limit = 1048576
	}
	if limit < 0 || len(data) > limit || !utf8.ValidString(data) {
		return nil, fmt.Errorf("invalid SSE data or byte limit exceeded")
	}
	for _, value := range []*string{options.Event, options.ID} {
		if value != nil && (len(*value) > 1024 || strings.ContainsAny(*value, "\r\n\x00") || !utf8.ValidString(*value)) {
			return nil, fmt.Errorf("invalid SSE event name or ID")
		}
	}
	if options.Retry != nil && *options.Retry < 0 {
		return nil, fmt.Errorf("SSE retry must be nonnegative")
	}
	var out strings.Builder
	if options.Event != nil {
		out.WriteString("event: " + *options.Event + "\n")
	}
	normalized := strings.ReplaceAll(strings.ReplaceAll(data, "\r\n", "\n"), "\r", "\n")
	for _, line := range strings.Split(normalized, "\n") {
		out.WriteString("data: " + line + "\n")
	}
	if options.ID != nil {
		out.WriteString("id: " + *options.ID + "\n")
	}
	if options.Retry != nil {
		out.WriteString("retry: " + strconv.FormatInt(*options.Retry, 10) + "\n")
	}
	out.WriteByte('\n')
	return []byte(out.String()), nil
}
